### ISSUE TITLE
Calls to an API defining error status code on routes with C# AutoRest client returns null instead of raising the associated error

### ISSUE THREAD

**Issue by ggirard07** [Azure/autorest#2239]
**Steps to reproduce**
1. Setup an API with a single controller with a single route with following attributes:
```
[SwaggerResponse(HttpStatusCode.OK)]
[SwaggerResponse(HttpStatusCode.NotFound)]
[SwaggerResponse(HttpStatusCode.Unauthorized)]
```
2. Ensures the route returns proper status code (`HttpStatusCode.NotFound` for a bad parameter or `HttpStatusCode.Unauthorized` if bearer token is missing).
3. Generate an AutoRest client in CSharp based on API json doc with option `-AddCredentials true`.
4. Call the route on the API controller with proper parameter in order to raise the `HttpStatusCode.NotFound` or `HttpStatusCode.Unauthorized`.

**Expected result**
AutoRest client must allow developer to know about the involved errors.

**Actual result**
Null is returned. No error is signaled by the AutoRest client.

**Remarks**
I am not sure if this is a bug, by design or I am missing something to get the expected error.
If it is by design, I don't understand why if an error code is documented on a route it means the developer does not have to know about it...

**Versions**
```
PS C:\Git\EngineeringWorkflow\Distech.EWT.MVC> autorest --list-installed
AutoRest Code Generator
(c) 2017 Microsoft Corporation. https://aka.ms/autorest
  Build Information
  Autorest Bootstrapper :  0.13.2
  NetCore framework :      1.0.4
  Latest Core Installed :  1.0.1-20170501-2300-nightly
  Requested Core Version : <none>
Installed versions of AutoRest :

  1.0.1-20170501-2300-nightly
```

**Comment by olydis** [comment]
Since AutoRest's job is converting Swagger into code, please provide the Swagger that causes the issue you describe. What do you mean with "Setup an API"? The code you posted (`[SwaggerResponse(HttpStatusCode.NotFound)]` and such) is nothing AutoRest processes.

**Comment by ggirard07** [comment]
@olydis I published the API definition in a [gist](https://gist.github.com/ggirard07/4ae90f5b62a692eeb97e9d382d075d71).

Here an extracted of the code from the API which produce this JSON:
```
[RoutePrefix("api/v1/projects")]
public class ProjectsController
    : ApiController
{
    private IDataService m_ProjectService;

    public ProjectsController(IDataService service)
    {
        m_ProjectService = service;
    }

    [HttpGet]
    [Route("")]
    [SwaggerOperation("GetAllProjects")]
    [SwaggerResponse(HttpStatusCode.Unauthorized)]
    [AuthorizeScope(Helper.ReadAccountsScope)]
    public IQueryable<ProjectViewModel> GetProjects()
    {
        return m_ProjectService.Projects.ProjectTo<ProjectViewModel>();
    }
    
    [HttpGet]
    [Route("{id:int}", Name = "GetProjectById")]
    [ResponseType(typeof(ProjectViewModel))]
    [SwaggerOperation("GetProjectById")]
    [SwaggerResponse(HttpStatusCode.OK)]
    [SwaggerResponse(HttpStatusCode.NotFound)]
    [SwaggerResponse(HttpStatusCode.Unauthorized)]
    [AuthorizeScope(Helper.ReadAccountsScope)]
    public async Task<IHttpActionResult> GetProjectAsync(int id)
    {
        Project project = await m_ProjectService.GetProjectAsync(id);
        if (project == null)
            return NotFound();

        return Ok(AutoMapper.Mapper.Map<ProjectViewModel>(project));
    }
    
    ...
}
```

Depending whether I comment the `[SwaggerResponse(HttpStatusCode.NotFound)]` from my API controller, the `responses` section for the corresponding routes in the JSON contains, or not, the 404 definition. If it is uncommented on my controller and so defined in the JSON, then the client generated by AutoRest silents the 404 error.

**Code generated with the 404 defined in JSON:**
```
HttpStatusCode _statusCode = _httpResponse.StatusCode;
cancellationToken.ThrowIfCancellationRequested();
string _responseContent = null;
if ((int)_statusCode != 200 && (int)_statusCode != 401 && (int)_statusCode != 404)
{
    var ex = new HttpOperationException(string.Format("Operation returned an invalid status code '{0}'", _statusCode));
    if (_httpResponse.Content != null) {
        _responseContent = await _httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
    }
    else {
        _responseContent = string.Empty;
    }
    ex.Request = new HttpRequestMessageWrapper(_httpRequest, _requestContent);
    ex.Response = new HttpResponseMessageWrapper(_httpResponse, _responseContent);
    if (_shouldTrace)
    {
        ServiceClientTracing.Error(_invocationId, ex);
    }
    _httpRequest.Dispose();
    if (_httpResponse != null)
    {
        _httpResponse.Dispose();
    }
    throw ex;
}
```
**Code generated without the 404 defined in JSON:**
```
HttpStatusCode _statusCode = _httpResponse.StatusCode;
cancellationToken.ThrowIfCancellationRequested();
string _responseContent = null;
if ((int)_statusCode != 200 && (int)_statusCode != 401)
{
    var ex = new HttpOperationException(string.Format("Operation returned an invalid status code '{0}'", _statusCode));
    if (_httpResponse.Content != null) {
        _responseContent = await _httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
    }
    else {
        _responseContent = string.Empty;
    }
    ex.Request = new HttpRequestMessageWrapper(_httpRequest, _requestContent);
    ex.Response = new HttpResponseMessageWrapper(_httpResponse, _responseContent);
    if (_shouldTrace)
    {
        ServiceClientTracing.Error(_invocationId, ex);
    }
    _httpRequest.Dispose();
    if (_httpResponse != null)
    {
        _httpResponse.Dispose();
    }
    throw ex;
}
```

The `(int)_statusCode != 404` condition in the if statement in charge to handle errors which silent it is generated only if the `[SwaggerResponse(HttpStatusCode.NotFound)]` is defined on this precise route.

**Comment by olydis** [comment]
Whatever tool you use to generate Swagger from your API controller seems to have slightly different conventions regarding status codes: https://github.com/Azure/azure-rest-api-specs-pr/blob/master/documentation/creating-swagger.md#negative-responses

TL;DR
any status code documented explicitly in your Swagger is considered a "success" by AutoRest. This is arguably not the best way to deal with it and we plan to make that configurable in the future, but currently this is what AutoRest does and what most of its users rely on. So if possible, adjust the Swagger accordingly.

**Comment by olydis** [comment]
will close this issue as it is a known problem (tracked at least here: https://github.com/Azure/autorest/issues/1523)
